### PR TITLE
Add remapping for diagnostic output on latlong grids

### PR DESCRIPTION
**Pull request by jb-mackay**
Addresses #742 

- Introduces a lightweight `RegularLatLongMesh` struct describing the grid to record output on.
- Introduces `RLLRemap`, a remapping operator storing the interpolation matrix between a field and the lat/long mesh.

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: src/Meshes/latlong.jl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    lat::R
+    long::R
+end
+
+function RegularLatLongMesh(
+    domain,
+    nlat::Int,
+    nlong::Int,
+    lat_bounds = (-π / 2, π / 2),
+    long_bounds = (-π, π),
```
Comment (simonbyrne) [inline src/Meshes/latlong.jl:15]:
```suggestion
    lat_bounds = (-π / 2, π / 2),
    long_bounds = (-π, π),
```
we use degrees for bounds everywhere else

File: src/Meshes/latlong.jl [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+)
+    @assert 0 < (lat_bounds[2] - lat_bounds[1]) <= π
+    @assert 0 < (long_bounds[2] - long_bounds[1]) <= 2π
+
+    FT = typeof(domain.radius)
+    lat = range(ClosedInterval{FT}(lat_bounds[1], lat_bounds[2]), nlat) # lat [-π/2, π/2]
```
Comment (simonbyrne) [inline src/Meshes/latlong.jl:21]:
```suggestion
    lat = range(lat_bounds[1], lat_bounds[2], length=nlat)
```

File: src/Meshes/latlong.jl [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    @assert -180 <= long_bounds[1] < long_bounds[2] <= 180
+
+    FT = typeof(domain.radius)
+    lat = range(lat_bounds[1], lat_bounds[2], length = nlat)
+    if long_bounds[1] + 360 ≈ long_bounds[2] # periodic, long_bounds[2] "=" long_bounds[1]
+        long = range(
+            Interval{:closed, :open, FT}(long_bounds[1], long_bounds[2]),
+            nlong,
+        )
+    else
```
Comment (simonbyrne) [inline src/Meshes/latlong.jl:27]: what's the point of this branch?

Comment (jb-mackay) [reply src/Meshes/latlong.jl:27]: Periodicity - if the longitude range is set as (-180, 180) you don't want two points at the same spot, so it creates an open interval and the endpoint is not included in the range.

**Comment by simonbyrne** [comment]
This is a good start: can you add some basic test cases (e.g. interpolate the coordinate field of a cubed sphere to an RLL grid)?

Then need to add some more pieces: 
- NetCDF output
- remapping of extruded fields
- distributed support

**Comment by jb-mackay** [comment]
> This is a good start: can you add some basic test cases (e.g. interpolate the coordinate field of a cubed sphere to an RLL grid)?
> 
> Then need to add some more pieces:
> 
> * NetCDF output
> * remapping of extruded fields
> * distributed support

Can do - let's connect once some tests are in place.

File: src/Operators/remapping.jl [hunk 4]
```diff
@@ -290,3 +290,103 @@ function local_weights(space::AbstractSpace)
...
+    coords = map(coords, 1:length(coords)) do coord, i
+        (coord..., idx = i)
+    end
+    # get unique elements and the indices of the first associated coords
+    elems = map(x -> (elem = x.elem, idx = x.idx), unique(x -> x.elem, coords))
+
+    QS_s = Spaces.quadrature_style(source_space)
+    Nq_s = Quadratures.degrees_of_freedom(QS_s)
+    # rows - for each rll coord, cols - for each node in corresponding elem
+    op = spzeros(length(coords), Nq_s^2)
```
Comment (simonbyrne) [inline src/Operators/remapping.jl:336]: should be a dense matrix, also switch order of indices (as they will be memory aligned)

Comment (simonbyrne) [reply src/Operators/remapping.jl:336]: Actually, may be better to do it as `zeros(Nq_s, Nq_s, length(coords))`

File: src/Operators/remapping.jl [hunk 5]
```diff
@@ -290,3 +290,103 @@ function local_weights(space::AbstractSpace)
...
+    data = parent(source_field)
+    for (i, elem) in enumerate(R.elems)
+        elem_idx =
+            Topologies.localelemindex(axes(source_field).topology, elem.elem)
+        start_idx = elem.idx
+        last_idx =
+            i == length(R.elems) ? length(target_vec) : R.elems[i + 1].idx - 1
+        for j in start_idx:last_idx
+            # multiply rows of map by nodal values in corresponding element
+            target_vec[j] = R.map[j, :]' * vec(data[:, :, 1, elem_idx])
```
Comment (simonbyrne) [inline src/Operators/remapping.jl:375]:
```suggestion
            target_vec[j] = dot(view(R.map, :, :, j), view(data, :, :, 1, elem_idx))
```

File: src/Operators/remapping.jl [hunk 6]
```diff
@@ -290,3 +290,103 @@ function local_weights(space::AbstractSpace)
...
+
+    # store latlong coords and associated elements
+    coords = []
+    for lat in latitude
+        for long in longitude
+            coord = Geometry.LatLongPoint(lat, long)
+            elem = Meshes.containing_element(source_mesh, coord)
+            push!(coords, (coord = coord, elem = elem))
+        end
+    end
```
Comment (simonbyrne) [inline src/Operators/remapping.jl:322]: may be more efficient to do this as a comprehension:
```suggestion
    coords = [begin
            coord = Geometry.LatLongPoint(lat, long)
            elem = Meshes.containing_element(source_mesh, coord)
            (coord = coord, elem = elem)
        end for long in longitude, lat in latitude]
```

File: test/Operators/remapping.jl [hunk 7]
```diff
@@ -725,3 +725,29 @@ end
...
+    space = Spaces.SpectralElementSpace2D(topology, quad)
+    coords = Fields.coordinate_field(space)
+
+    rllmesh = Meshes.RegularLatLongMesh(spheredomain, 100, 100)
+    rllcoords = Meshes.rllcoords(rllmesh)
+
+    R = Operators.RLLRemap(rllmesh, space)
+
+    source_field = sind.(coords.long) .* cosd.(coords.lat)
+    remap_vec = Operators.remap(R, source_field)
```
Comment (simonbyrne) [inline test/Operators/remapping.jl:748]: add a test for `remap!`

File: test/Operators/remapping.jl [hunk 8]
```diff
@@ -725,3 +725,33 @@ end
...
+
+@testset "RegularLatLong Interpolation" begin
+    radius = FT(3)
+    ne = 10
+    Nq = 4
+    spheredomain = Domains.SphereDomain(radius)
+    # each of the 6 faces has 4x4 elements with 4x4 nodes per elem
+    # 96 elems of 16 nodes each
+    mesh = Meshes.EquiangularCubedSphere(spheredomain, ne)
+    topology = Topologies.Topology2D(mesh)
```
Comment (simonbyrne) [inline test/Operators/remapping.jl:737]: Can you add a test for a space filling curve as well?

**Pull request closed without merging**